fuzz-test-suite/fuzzamericanoption.cpp

```diff
@@ -0,0 +1,132 @@
+        Option::Type type;
+        Real strike;
+        Real s;       // spot
+        Rate q;       // dividend
+        Rate r;       // risk-free rate
+        Time t;       // time to maturity
+        Volatility v; // volatility
+    };
+
+}
+
+AmericanOptionData fuzzedAmericanOptionData(FuzzedDataProvider& fdp) {
+    return AmericanOptionData{
+        .type = fdp.PickValueInArray({Option::Type::Put, Option::Type::Call}),
+        .strike = fdp.ConsumeFloatingPoint<Real>(),
+        .s = fdp.ConsumeFloatingPointInRange<Real>(0.0, std::numeric_limits<Real>::max()),
+        .q = fdp.ConsumeProbability<Rate>(),
+        .r = fdp.ConsumeProbability<Rate>(),
+        .t = fdp.ConsumeFloatingPointInRange<Time>(0.0, std::numeric_limits<Real>::max()),
+        // TODO(silvergasp): Confirm this should be between 0.0-1.0
+        .v = fdp.ConsumeProbability<Volatility>(),
+    };
+}
+
+extern "C" int LLVMFuzzerTestOneInput(const uint8_t* data, size_t size) {
+    FuzzedDataProvider fdp(data, size);
+    // Ensure that settings are reset between each fuzzing iteration.
+    SavedSettings saved_settings;
+
+    constexpr size_t kMaxValues = 1024;
+    auto length = fdp.ConsumeIntegralInRange<size_t>(0, kMaxValues);
+    std::vector<AmericanOptionData> values;
+    values.reserve(length);
+    for (size_t i = 0; i < length; i++) {
+        values.push_back(fuzzedAmericanOptionData(fdp));
+    }
+
+    Date today(2, March, 2020);
+    DayCounter dc = Actual360();
+    ext::shared_ptr<SimpleQuote> spot(new SimpleQuote(0.0));
+    ext::shared_ptr<SimpleQuote> qRate(new SimpleQuote(0.0));
+
+    ext::shared_ptr<YieldTermStructure> qTS =
+        ext::shared_ptr<YieldTermStructure>(new FlatForward(today, Handle<Quote>(qRate), dc));
+    ext::shared_ptr<SimpleQuote> rRate(new SimpleQuote(0.0));
+    ext::shared_ptr<YieldTermStructure> rTS =
+        ext::shared_ptr<YieldTermStructure>(new FlatForward(today, Handle<Quote>(rRate), dc));
+    ext::shared_ptr<SimpleQuote> vol(new SimpleQuote(0.0));
+    ext::shared_ptr<BlackVolTermStructure> volTS = ext::shared_ptr<BlackVolTermStructure>(
+        new BlackConstantVol(today, NullCalendar(), Handle<Quote>(vol), dc));
+
+    for (auto& value : values) {
+
+        ext::shared_ptr<StrikedTypePayoff> payoff(new PlainVanillaPayoff(value.type, value.strike));
+        Date exDate = today + Integer(std::lround(365*value.t));
+        ext::shared_ptr<Exercise> exercise(new AmericanExercise(today, exDate));
+
+        spot->setValue(value.s);
+        qRate->setValue(value.q);
+        rRate->setValue(value.r);
+        vol->setValue(value.v);
+
+        ext::shared_ptr<BlackScholesMertonProcess> stochProcess(new BlackScholesMertonProcess(
+            Handle<Quote>(spot), Handle<YieldTermStructure>(qTS), Handle<YieldTermStructure>(rTS),
+            Handle<BlackVolTermStructure>(volTS)));
+
+        ext::shared_ptr<PricingEngine> engine(
+            new BaroneAdesiWhaleyApproximationEngine(stochProcess));
```

Comment:
This part can go out of the loop if you want to save some cycles

Reply:
Thanks, done.